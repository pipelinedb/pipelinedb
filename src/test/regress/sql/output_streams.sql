
CREATE CONTINUOUS VIEW os0 AS SELECT COUNT(*) FROM os_stream;
CREATE CONTINUOUS VIEW os0_output AS SELECT (old).count AS old_count, (new).count AS new_count FROM os0_osrel;

INSERT INTO os_stream (x) VALUES (0);
SELECT pg_sleep(2);

SELECT * FROM os0_output ORDER BY old_count, new_count;

INSERT INTO os_stream (x) VALUES (0);
SELECT pg_sleep(2);

SELECT * FROM os0_output ORDER BY old_count, new_count;

INSERT INTO os_stream (x) VALUES (0);
SELECT pg_sleep(2);

SELECT * FROM os0_output ORDER BY old_count, new_count;

INSERT INTO os_stream (x) VALUES (0);
SELECT pg_sleep(2);

SELECT * FROM os0_output ORDER BY old_count, new_count;

-- We shouldn't be able to drop this because os0_output depends on it now
DROP CONTINUOUS VIEW os0;

DROP CONTINUOUS VIEW os0_output;
DROP CONTINUOUS VIEW os0;

CREATE CONTINUOUS VIEW os1 AS SELECT abs(x::integer) AS g, COUNT(*), sum(x)
FROM os_stream GROUP BY g;

CREATE CONTINUOUS VIEW os1_output AS SELECT
  (new).g,
  (old).count AS old_count, (new).count AS new_count,
  (old).sum AS old_sum, (new).sum AS new_sum
FROM os1_osrel;

INSERT INTO os_stream (x) VALUES (10);
SELECT pg_sleep(2);

SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;

INSERT INTO os_stream (x) VALUES (20);
SELECT pg_sleep(2);

SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;

INSERT INTO os_stream (x) VALUES (30);
SELECT pg_sleep(2);

SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;

INSERT INTO os_stream (x) VALUES (40);
SELECT pg_sleep(2);

INSERT INTO os_stream (x) VALUES (-40);
SELECT pg_sleep(2);

SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;

INSERT INTO os_stream (x) VALUES (-30);
SELECT pg_sleep(2);

SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;

INSERT INTO os_stream (x) VALUES (-20);
SELECT pg_sleep(2);

SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;

INSERT INTO os_stream (x) VALUES (-10);
SELECT pg_sleep(2);

SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;

DROP CONTINUOUS VIEW os1 CASCADE;

-- Verify SW ticking into output streams
CREATE CONTINUOUS VIEW os2 WITH (max_age = '10 seconds') AS
SELECT x::integer, COUNT(*)
FROM os_stream GROUP BY x;

CREATE CONTINUOUS VIEW os2_output AS SELECT
  arrival_timestamp,
  CASE WHEN old IS NULL THEN (new).x ELSE (old).x END AS x,
  old, new
FROM os2_osrel;

INSERT INTO os_stream (x) VALUES (0);
SELECT pg_sleep(2);

SELECT x, old, new FROM os2_output ORDER BY x, old, new;

INSERT INTO os_stream (x) VALUES (1);
SELECT pg_sleep(2);

SELECT x, old, new FROM os2_output ORDER BY x, old, new;

INSERT INTO os_stream (x) VALUES (2);
SELECT pg_sleep(2);

SELECT x, old, new FROM os2_output ORDER BY x, old, new;

INSERT INTO os_stream (x) VALUES (3);
SELECT pg_sleep(20);

SELECT x, old, new FROM os2_output ORDER BY x, old, new;

DROP CONTINUOUS VIEW os2 CASCADE;

-- Stream-table joins on output streams
CREATE TABLE os_t0 (x integer, y integer);
INSERT INTO os_t0 (x, y) VALUES (0, 42);

CREATE CONTINUOUS VIEW os3 AS SELECT x::integer, COUNT(*)
FROM os_stream GROUP BY x;

CREATE CONTINUOUS VIEW os3_output AS SELECT
  CASE WHEN (s.old) IS NULL THEN (s.new).x ELSE (s.old).x END AS x,
  s.new, t.y
  FROM os3_osrel s JOIN os_t0 t ON x = t.x;

INSERT INTO os_stream (x) VALUES (0);
SELECT pg_sleep(2);

SELECT x, new, y FROM os3_output ORDER BY x, new, y;

INSERT INTO os_stream (x) VALUES (0);
SELECT pg_sleep(2);

SELECT x, new, y FROM os3_output ORDER BY x, new, y;

INSERT INTO os_stream (x) VALUES (0);
SELECT pg_sleep(2);

SELECT x, new, y FROM os3_output ORDER BY x, new, y;

DROP TABLE os_t0;
DROP CONTINUOUS VIEW os3 CASCADE;

-- Final functions should be applied to output stream tuples where necessary
CREATE CONTINUOUS VIEW os3 AS SELECT x::integer, avg(y::integer), count(distinct z::integer)
FROM os_stream GROUP BY x;

CREATE CONTINUOUS VIEW os3_output AS SELECT
  arrival_timestamp,
  CASE WHEN old IS NULL THEN (new).x ELSE (old).x END AS x,
  old, new
FROM os3_osrel;

INSERT INTO os_stream (x, y, z) VALUES (0, 2, 0);
SELECT pg_sleep(2);

SELECT x, old, new FROM os3_output ORDER BY x, old, new;

INSERT INTO os_stream (x, y, z) VALUES (0, 4, 1);
SELECT pg_sleep(2);

SELECT x, old, new FROM os3_output ORDER BY x, old, new;

INSERT INTO os_stream (x, y, z) VALUES (1, 8, 2);
SELECT pg_sleep(2);

SELECT x, old, new FROM os3_output ORDER BY x, old, new;

INSERT INTO os_stream (x, y, z) VALUES (1, 16, 2);
SELECT pg_sleep(2);

SELECT x, old, new FROM os3_output ORDER BY x, old, new;

DROP CONTINUOUS VIEW os3 CASCADE;

