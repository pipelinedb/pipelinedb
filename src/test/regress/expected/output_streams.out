CREATE STREAM os_stream (x int, y int, z int);
CREATE CONTINUOUS VIEW os0 AS SELECT COUNT(*) FROM os_stream;
CREATE CONTINUOUS VIEW os0_output AS SELECT (old).count AS old_count, (new).count AS new_count FROM output_of('os0');
INSERT INTO os_stream (x) VALUES (0);
SELECT * FROM os0_output ORDER BY old_count, new_count;
 old_count | new_count 
-----------+-----------
           |         1
(1 row)

INSERT INTO os_stream (x) VALUES (0);
SELECT * FROM os0_output ORDER BY old_count, new_count;
 old_count | new_count 
-----------+-----------
         1 |         2
           |         1
(2 rows)

INSERT INTO os_stream (x) VALUES (0);
SELECT * FROM os0_output ORDER BY old_count, new_count;
 old_count | new_count 
-----------+-----------
         1 |         2
         2 |         3
           |         1
(3 rows)

INSERT INTO os_stream (x) VALUES (0);
SELECT * FROM os0_output ORDER BY old_count, new_count;
 old_count | new_count 
-----------+-----------
         1 |         2
         2 |         3
         3 |         4
           |         1
(4 rows)

-- We shouldn't be able to drop this because os0_output depends on it now
DROP CONTINUOUS VIEW os0;
ERROR:  cannot drop continuous view os0 because other objects depend on it
DETAIL:  continuous view os0_output depends on stream os0_osrel
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
DROP CONTINUOUS VIEW os0_output;
DROP CONTINUOUS VIEW os0;
CREATE CONTINUOUS VIEW os1 AS SELECT abs(x::integer) AS g, COUNT(*), sum(x)
FROM os_stream GROUP BY g;
CREATE CONTINUOUS VIEW os1_output AS SELECT
  (new).g,
  (old).count AS old_count, (new).count AS new_count,
  (old).sum AS old_sum, (new).sum AS new_sum
FROM output_of('os1');
INSERT INTO os_stream (x) VALUES (10);
SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;
 g  | old_count | new_count | old_sum | new_sum 
----+-----------+-----------+---------+---------
 10 |           |         1 |         |      10
(1 row)

INSERT INTO os_stream (x) VALUES (20);
SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;
 g  | old_count | new_count | old_sum | new_sum 
----+-----------+-----------+---------+---------
 10 |           |         1 |         |      10
 20 |           |         1 |         |      20
(2 rows)

INSERT INTO os_stream (x) VALUES (30);
SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;
 g  | old_count | new_count | old_sum | new_sum 
----+-----------+-----------+---------+---------
 10 |           |         1 |         |      10
 20 |           |         1 |         |      20
 30 |           |         1 |         |      30
(3 rows)

INSERT INTO os_stream (x) VALUES (40);
INSERT INTO os_stream (x) VALUES (-40);
SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;
 g  | old_count | new_count | old_sum | new_sum 
----+-----------+-----------+---------+---------
 10 |           |         1 |         |      10
 20 |           |         1 |         |      20
 30 |           |         1 |         |      30
 40 |         1 |         2 |      40 |       0
 40 |           |         1 |         |      40
(5 rows)

INSERT INTO os_stream (x) VALUES (-30);
SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;
 g  | old_count | new_count | old_sum | new_sum 
----+-----------+-----------+---------+---------
 10 |           |         1 |         |      10
 20 |           |         1 |         |      20
 30 |         1 |         2 |      30 |       0
 30 |           |         1 |         |      30
 40 |         1 |         2 |      40 |       0
 40 |           |         1 |         |      40
(6 rows)

INSERT INTO os_stream (x) VALUES (-20);
SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;
 g  | old_count | new_count | old_sum | new_sum 
----+-----------+-----------+---------+---------
 10 |           |         1 |         |      10
 20 |         1 |         2 |      20 |       0
 20 |           |         1 |         |      20
 30 |         1 |         2 |      30 |       0
 30 |           |         1 |         |      30
 40 |         1 |         2 |      40 |       0
 40 |           |         1 |         |      40
(7 rows)

INSERT INTO os_stream (x) VALUES (-10);
SELECT * FROM os1_output ORDER BY g, old_count, new_count, old_sum, new_sum;
 g  | old_count | new_count | old_sum | new_sum 
----+-----------+-----------+---------+---------
 10 |         1 |         2 |      10 |       0
 10 |           |         1 |         |      10
 20 |         1 |         2 |      20 |       0
 20 |           |         1 |         |      20
 30 |         1 |         2 |      30 |       0
 30 |           |         1 |         |      30
 40 |         1 |         2 |      40 |       0
 40 |           |         1 |         |      40
(8 rows)

DROP CONTINUOUS VIEW os1 CASCADE;
NOTICE:  drop cascades to continuous view os1_output
-- Verify SW ticking into output streams
CREATE CONTINUOUS VIEW os2 WITH (sw = '10 seconds') AS
SELECT x::integer, COUNT(*)
FROM os_stream GROUP BY x;
CREATE CONTINUOUS VIEW os2_output AS SELECT
  arrival_timestamp,
  CASE WHEN old IS NULL THEN (new).x ELSE (old).x END AS x,
  old, new
FROM output_of('os2');
INSERT INTO os_stream (x) VALUES (0);
SELECT x, old, new FROM os2_output ORDER BY x, old, new;
 x | old |  new  
---+-----+-------
 0 |     | (0,1)
(1 row)

INSERT INTO os_stream (x) VALUES (1);
SELECT x, old, new FROM os2_output ORDER BY x, old, new;
 x | old |  new  
---+-----+-------
 0 |     | (0,1)
 1 |     | (1,1)
(2 rows)

INSERT INTO os_stream (x) VALUES (2);
SELECT x, old, new FROM os2_output ORDER BY x, old, new;
 x | old |  new  
---+-----+-------
 0 |     | (0,1)
 1 |     | (1,1)
 2 |     | (2,1)
(3 rows)

INSERT INTO os_stream (x) VALUES (3);
SELECT pg_sleep(12);
 pg_sleep 
----------
 
(1 row)

SELECT x, old, new FROM os2_output ORDER BY x, old, new;
 x |  old  |  new  
---+-------+-------
 0 | (0,1) | 
 0 |       | (0,1)
 1 | (1,1) | 
 1 |       | (1,1)
 2 | (2,1) | 
 2 |       | (2,1)
 3 | (3,1) | 
 3 |       | (3,1)
(8 rows)

DROP CONTINUOUS VIEW os2 CASCADE;
NOTICE:  drop cascades to continuous view os2_output
-- Stream-table joins on output streams
CREATE TABLE os_t0 (x integer, y integer);
INSERT INTO os_t0 (x, y) VALUES (0, 42);
CREATE CONTINUOUS VIEW os3 AS SELECT x::integer, COUNT(*)
FROM os_stream GROUP BY x;
CREATE CONTINUOUS VIEW os3_output AS SELECT
  CASE WHEN (old) IS NULL THEN (new).x ELSE (old).x END AS x,
  new, t.y
  FROM output_of('os3') JOIN os_t0 t ON x = t.x;
NOTICE:  consider creating an index on t.x for improved stream-table join performance
INSERT INTO os_stream (x) VALUES (0);
SELECT x, new, y FROM os3_output ORDER BY x, new, y;
 x |   new   | y  
---+---------+----
 0 | (0,1,1) | 42
(1 row)

INSERT INTO os_stream (x) VALUES (0);
SELECT x, new, y FROM os3_output ORDER BY x, new, y;
 x |   new   | y  
---+---------+----
 0 | (0,1,1) | 42
 0 | (0,2,1) | 42
(2 rows)

INSERT INTO os_stream (x) VALUES (0);
SELECT x, new, y FROM os3_output ORDER BY x, new, y;
 x |   new   | y  
---+---------+----
 0 | (0,1,1) | 42
 0 | (0,2,1) | 42
 0 | (0,3,1) | 42
(3 rows)

DROP TABLE os_t0;
ERROR:  cannot drop table os_t0 because other objects depend on it
DETAIL:  continuous view os3_output depends on table os_t0
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
DROP CONTINUOUS VIEW os3 CASCADE;
NOTICE:  drop cascades to continuous view os3_output
-- Final functions should be applied to output stream tuples where necessary
CREATE CONTINUOUS VIEW os3 AS SELECT x::integer, avg(y::integer), count(distinct z::integer)
FROM os_stream GROUP BY x;
CREATE CONTINUOUS VIEW os3_output AS SELECT
  arrival_timestamp,
  CASE WHEN old IS NULL THEN (new).x ELSE (old).x END AS x,
  old, new
FROM output_of('os3');
INSERT INTO os_stream (x, y, z) VALUES (0, 2, 0);
SELECT x, old, new FROM os3_output ORDER BY x, old, new;
 x | old |           new            
---+-----+--------------------------
 0 |     | (0,2.0000000000000000,1)
(1 row)

INSERT INTO os_stream (x, y, z) VALUES (0, 4, 1);
SELECT x, old, new FROM os3_output ORDER BY x, old, new;
 x |           old            |           new            
---+--------------------------+--------------------------
 0 | (0,2.0000000000000000,1) | (0,3.0000000000000000,2)
 0 |                          | (0,2.0000000000000000,1)
(2 rows)

INSERT INTO os_stream (x, y, z) VALUES (1, 8, 2);
SELECT x, old, new FROM os3_output ORDER BY x, old, new;
 x |           old            |           new            
---+--------------------------+--------------------------
 0 | (0,2.0000000000000000,1) | (0,3.0000000000000000,2)
 0 |                          | (0,2.0000000000000000,1)
 1 |                          | (1,8.0000000000000000,1)
(3 rows)

INSERT INTO os_stream (x, y, z) VALUES (1, 16, 2);
SELECT x, old, new FROM os3_output ORDER BY x, old, new;
 x |           old            |            new            
---+--------------------------+---------------------------
 0 | (0,2.0000000000000000,1) | (0,3.0000000000000000,2)
 0 |                          | (0,2.0000000000000000,1)
 1 | (1,8.0000000000000000,1) | (1,12.0000000000000000,1)
 1 |                          | (1,8.0000000000000000,1)
(4 rows)

DROP CONTINUOUS VIEW os3 CASCADE;
NOTICE:  drop cascades to continuous view os3_output
-- Verify that transforms write to output streams
CREATE CONTINUOUS TRANSFORM os_xform AS SELECT x, y FROM os_stream;
CREATE CONTINUOUS VIEW os4 AS SELECT x, y FROM output_of('os_xform');
INSERT INTO os_stream (x, y) VALUES (7, 7);
INSERT INTO os_stream (x, y) VALUES (8, 8);
INSERT INTO os_stream (x, y) VALUES (9, 9);
SELECT * FROM os4 ORDER BY x;
 x | y 
---+---
 7 | 7
 8 | 8
 9 | 9
(3 rows)

DROP STREAM os_stream CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to continuous transform os_xform
drop cascades to continuous view os4
